### PR TITLE
[SYCL] Set preprocessor macros in integration header

### PR DESCRIPTION
**Pull request by elizabethandrews**
Predefines which need to be set for custom host compilation
must be defined in integration header.

Signed-off-by: Elizabeth Andrews <elizabeth.andrews@intel.com>

### CODE REVIEW

File: clang/include/clang/Basic/Version.h [hunk 1]
```diff
@@ -56,6 +58,9 @@ namespace clang {
   /// for use in the CPP __VERSION__ macro, which includes the clang version
   /// number, the repository version, and the vendor tag.
   std::string getClangFullCPPVersion();
+
+  llvm::SmallVector<std::pair<llvm::StringRef, llvm::StringRef>, 2>
+  getSYCLVersionMacros(const LangOptions &LangOpts);
```
Comment (smanna12) [inline clang/include/clang/Basic/Version.h:63]: Can you add a comment here (example line#57)?

Comment (elizabethandrews) [reply clang/include/clang/Basic/Version.h]: Thanks for the review! Done!

Comment (smanna12) [reply clang/include/clang/Basic/Version.h]: Thanks Elizabeth.

**Review by smanna12 — APPROVED** [review]
Changes look good to me. @premanandrao?

**Review by premanandrao — APPROVED** [review]
Changes LGTM

**Comment by dm-vodopyanov** [comment]
@elizabethandrews, can you please address failures in buildbot/sycl-win-x64-pr?

**Comment by bader** [comment]
> @elizabethandrews, can you please address failures in buildbot/sycl-win-x64-pr?

ping.

**Comment by elizabethandrews** [comment]
> > @elizabethandrews, can you please address failures in buildbot/sycl-win-x64-pr?
> 
> ping.

Taking a look now. I'm not sure why adding a few defines in integration header affects compilation of headers. I've updated the workspace to see if that fixes it. If not, I guess I will have to build a windows workspace and test

**Review by bader — CHANGES_REQUESTED** [review]
Please, fix buildbot/sycl-win-x64-pr failures.

File: sycl/include/CL/sycl/handler.hpp [hunk 2]
```diff
@@ -1062,7 +1062,7 @@ class __SYCL_EXPORT handler {
   }
 
 #ifdef SYCL_LANGUAGE_VERSION
-#define __SYCL_KERNEL_ATTR__ __attribute__((sycl_kernel))
+#define __SYCL_KERNEL_ATTR__ [[clang::sycl_kernel]]
```
Comment (elizabethandrews) [inline sycl/include/CL/sycl/handler.hpp:1065]: Spelling change is required to avoid compilation error when using MSVC custom host compiler

**Review by steffenlarsen — APPROVED** [review]
Runtime changes LGTM.

**Review by smanna12 — APPROVED** [review]

File: clang/lib/Basic/Version.cpp [hunk 3]
```diff
@@ -123,4 +123,13 @@ std::string getClangFullCPPVersion() {
   return buf;
 }
 
+llvm::SmallVector<std::pair<llvm::StringRef, llvm::StringRef>, 2>
+getSYCLVersionMacros(const LangOptions &LangOpts) {
+  if (LangOpts.getSYCLVersion() == LangOptions::SYCL_2017)
+    return {{"CL_SYCL_LANGUAGE_VERSION", "121"},
+            {"SYCL_LANGUAGE_VERSION", "201707"}};
+  else if (LangOpts.getSYCLVersion() == LangOptions::SYCL_2020)
```
Comment (bader) [inline clang/lib/Basic/Version.cpp:131]:
```suggestion
  if (LangOpts.getSYCLVersion() == LangOptions::SYCL_2020)
```
`else` is redundant here.
https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

**Review by bader — APPROVED** [review]

File: clang/lib/Basic/Version.cpp (hunk 3 above)
Comment (elizabethandrews) [reply clang/lib/Basic/Version.cpp]: Thanks! Done!

**Review by smanna12 — APPROVED** [review]

**Review by steffenlarsen — APPROVED** [review]

**Pull request merged**